src/main/java/io/vertx/core/net/impl/TCPServerBase.java

```diff
@@ -120,102 +117,159 @@ private synchronized io.netty.util.concurrent.Future<Channel> listen(SocketAddre
     this.eventLoop = context.nettyEventLoop();
 
     SocketAddress bindAddress;
+    actualPort = localAddress.port();
+    String hostOrPath = localAddress.isInetSocket() ? localAddress.host() : localAddress.path();
+    boolean shared;
+    ServerID id;
+
+    if (actualPort > 0 || localAddress.isDomainSocket()) {
+      id = new ServerID(actualPort, hostOrPath);
+      shared = true;
+      bindAddress = localAddress;
+    } else {
+      if (actualPort < 0) {
+        id = new ServerID(actualPort, hostOrPath + "/" + -actualPort);
+        shared = true;
+        bindAddress = SocketAddress.inetSocketAddress(0, localAddress.host());
+      } else {
+        id = new ServerID(actualPort, hostOrPath);
+        shared = false;
+        bindAddress = localAddress;
+      }
+    }
+
     Map<ServerID, TCPServerBase> sharedNetServers = vertx.sharedTCPServers((Class<TCPServerBase>) getClass());
-    synchronized (sharedNetServers) {
-      actualPort = localAddress.port();
-      String hostOrPath = localAddress.isInetSocket() ? localAddress.host() : localAddress.path();
+
+    if (shared) {
       TCPServerBase main;
-      boolean shared;
-      ServerID id;
-      if (actualPort > 0 || localAddress.isDomainSocket()) {
-        id = new ServerID(actualPort, hostOrPath);
+
+      synchronized (sharedNetServers) {
         main = sharedNetServers.get(id);
-        shared = true;
-        bindAddress = localAddress;
-      } else {
-        if (actualPort < 0) {
-          id = new ServerID(actualPort, hostOrPath + "/" + -actualPort);
-          main = sharedNetServers.get(id);
-          shared = true;
-          bindAddress = SocketAddress.inetSocketAddress(0, localAddress.host());
-        } else {
-          id = new ServerID(actualPort, hostOrPath);
-          main = null;
-          shared = false;
-          bindAddress = localAddress;
+
+        if (main != null) {
+          try {
+            main.initialization.await();
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            return vertx.getAcceptorEventLoopGroup().next().newFailedFuture(e);
+          }
```

Comment:
@vietj , since this effectively blocks the thread waiting for the initialization of the other server, I think it may also need to run with the blocking thread pool. Wdyt?

Reply:
The PR is updated to await initialization via a blocking thread.